File: projeto/frontend/site/vox/src/app/services/subscription.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Subscription {
  id: number;
  userId: number;
  type: string;
  targetId: number | null;
}

@Injectable({ providedIn: 'root' })
export class SubscriptionService {
  private readonly API_URL = 'http://localhost:8080';

  constructor(private http: HttpClient) {}

  getSubscriptions(): Observable<Subscription[]> {
    return this.http.get<Subscription[]>(`${this.API_URL}/api/subscriptions`);
  }

  subscribeAllProjects(): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/subscriptions/all-projects`, null);
  }

  unsubscribeAllProjects(): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/api/subscriptions/all-projects`);
  }

  subscribeAllIssues(): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/subscriptions/all-issues`, null);
  }

  unsubscribeAllIssues(): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/api/subscriptions/all-issues`);
  }

  subscribeProject(projectId: number): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/subscriptions/projects/${projectId}`, null);
  }

  unsubscribeProject(projectId: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/api/subscriptions/projects/${projectId}`);
  }

  subscribeIssue(issueId: number): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/subscriptions/issues/${issueId}`, null);
  }

  unsubscribeIssue(issueId: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/api/subscriptions/issues/${issueId}`);
  }

  subscribeCategory(categoryId: number): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/subscriptions/categories/${categoryId}`, null);
  }

  unsubscribeCategory(categoryId: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/api/subscriptions/categories/${categoryId}`);
  }

  subscribeCouncilor(councilorId: number): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/api/subscriptions/councilors/${councilorId}`, null);
  }

  unsubscribeCouncilor(councilorId: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/api/subscriptions/councilors/${councilorId}`);
  }
}

File: projeto/frontend/app/app_vox/lib/services/subscription_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_client.dart';
import '../models/subscription.dart';

class SubscriptionService {
  Future<List<Subscription>> getSubscriptions() async {
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/api/subscriptions'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final list = jsonDecode(response.body) as List;
    return list
        .map((e) => Subscription.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<void> subscribeAllProjects() => _post('/api/subscriptions/all-projects');
  Future<void> unsubscribeAllProjects() => _delete('/api/subscriptions/all-projects');

  Future<void> subscribeAllIssues() => _post('/api/subscriptions/all-issues');
  Future<void> unsubscribeAllIssues() => _delete('/api/subscriptions/all-issues');

  Future<void> subscribeProject(int projectId) =>
      _post('/api/subscriptions/projects/$projectId');
  Future<void> unsubscribeProject(int projectId) =>
      _delete('/api/subscriptions/projects/$projectId');

  Future<void> subscribeIssue(int issueId) =>
      _post('/api/subscriptions/issues/$issueId');
  Future<void> unsubscribeIssue(int issueId) =>
      _delete('/api/subscriptions/issues/$issueId');

  Future<void> subscribeCategory(int categoryId) =>
      _post('/api/subscriptions/categories/$categoryId');
  Future<void> unsubscribeCategory(int categoryId) =>
      _delete('/api/subscriptions/categories/$categoryId');

  Future<void> subscribeCouncilor(int councilorId) =>
      _post('/api/subscriptions/councilors/$councilorId');
  Future<void> unsubscribeCouncilor(int councilorId) =>
      _delete('/api/subscriptions/councilors/$councilorId');

  Future<void> _post(String path) async {
    final response = await http.post(
      Uri.parse('${ApiClient.baseUrl}$path'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
  }

  Future<void> _delete(String path) async {
    final response = await http.delete(
      Uri.parse('${ApiClient.baseUrl}$path'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
  }
}
